src/contracts/SwapperV2.sol

```diff
@@ -0,0 +1,269 @@
+  }
+
+  /*///////////////////////////////////////////////////////////////
+                            External Functions
+  //////////////////////////////////////////////////////////////*/
+
+  /// @inheritdoc ISwapperV2
+  function deposit(uint256 _amount) external payable {
+    if (_amount == 0) revert Swapper_InvalidAmount();
+
+    // Ensure funds are deposited to the swapper contract
+    if (DEPOSITED_TOKEN == address(0)) {
+      if (msg.value != _amount) revert Swapper_AmountMismatch();
+    } else {
+      if (msg.value != 0) revert Swapper_AmountMismatch();
+      IERC20(DEPOSITED_TOKEN).safeTransferFrom(msg.sender, address(this), _amount);
+    }
+
+    // Add the deposit to the user's deposits by swap index
+    _userDeposits[msg.sender].push(Deposits({amount: _amount, swapIndex: _swapIndex}));
+
+    emit TokensDeposited(msg.sender, _amount);
+  }
+
+  /// @inheritdoc ISwapperV2
+  function swap() external payable override onlyOwner nonReentrant {
+    address[] memory path = _constructTokenPath();
+
+    uint256 _totalDeposited = _getTokenBalance(DEPOSITED_TOKEN);
+    uint256[] memory amounts;
+
+    if (DEPOSITED_TOKEN == address(0)) {
+      // swapping ETH -> token
+      amounts = ROUTER.swapExactETHForTokens{value: _totalDeposited}(0, path, address(this), block.timestamp);
+    } else {
+      // Approve the router to spend the deposited tokens
+      IERC20(DEPOSITED_TOKEN).approve(address(ROUTER), _totalDeposited);
+
+      if (SWAPPED_TOKEN == address(0)) {
+        // Swapping token -> ETH
+        amounts = ROUTER.swapExactTokensForETH(
+          _totalDeposited,
+          0, // Note: min amount out should be passed in as arg
+          path,
+          address(this),
+          block.timestamp
+        );
+      } else {
+        // Swapping token -> token
+        amounts = ROUTER.swapExactTokensForTokens(
+          _totalDeposited,
+          0, // Note: min amount out should be passed in as arg
+          path,
+          address(this),
+          block.timestamp
+        );
+      }
+    }
+
+    uint256 _totalSwapped = amounts[1];
+
+    // Store the swap rate information
+    _swaps[_swapIndex] = SwapRateInfo({totalDeposited: _totalDeposited, totalSwapped: _totalSwapped});
+
+    // Increment the swap index
+    _swapIndex++;
+
+    emit TokensSwapped(_totalDeposited, _totalSwapped);
+  }
+
+  /// @inheritdoc ISwapperV2
+  function withdrawDeposit() external nonReentrant {
+    Deposits[] storage _deposits = _userDeposits[msg.sender];
+    uint256 _withdrawableAmount = 0;
+
+    uint256 i = 0;
+    while (i < _deposits.length) {
+      if (_deposits[i].swapIndex >= _swapIndex) {
+        _withdrawableAmount += _deposits[i].amount;
+
+        // Remove the element by swapping with the last and popping
+        _deposits[i] = _deposits[_deposits.length - 1];
+        _deposits.pop();
+        // Do not increment i since the new item at i needs to be checked
+      } else {
+          i++;
+      }
+    }
```

Comment:
Iterating through and modifying a dynamic array in storage (_deposits) is expensive. For users with large deposit lists, this could lead to out-of-gas error.